### PR TITLE
feat: security section expansion

### PR DESCRIPTION
**Pull request by mdt2**
## Description
Expand the Security section to encompass what we've been learning through workshops with Bill and the Secure Dev Practices Task Force.

## Resources
- The contents of the PR come from this peer-reviewed, Naomi-reviewed document: https://docs.google.com/document/d/1Kp2UfOTq2R_eidnW3dBBB-Ottu0n9538PD1hNPGJW9g/edit?usp=sharing
- The contents of the document above are derived from the checklist Bill provided us with: https://docs.google.com/document/d/1tQpVG4X4nvxfCGRz4hqcg1NN2nWxvnZVEwTpv_WJfoI/edit

## To Validate
- [ ] Compare the contents of the PR to [Bill's document](https://docs.google.com/document/d/1tQpVG4X4nvxfCGRz4hqcg1NN2nWxvnZVEwTpv_WJfoI/edit). Is everything covered?
- [ ] Does the structure of the sections make sense?
- [ ] Does the content of the sections make sense?
- [ ] Do you see any places where examples or links may be helpful?

### CODE REVIEW

File: security/README.md [hunk 1]
```diff
@@ -21,35 +21,22 @@ assets of Sparkbox and its clients.  When in doubt about any matter related to
...
-screenshots], or Android. Reach out to the #security Slack channel with 
-questions or challenges.
+- **[Security Principles](security_principles)**
+- **Frontend:**
+  - **[JavaScript](frontend/javascript)**
+  - **[Exception Management](frontend/exception_management)**
+- **Backend:**
+  - **[Access Control](backend/access_control)**
+  - **[File Management](backend/file_management)**
+  - **[Deployment](backend/deployment)**
```
Comment (cromwellryan) [inline security/README.md:35]: Can _Deployment_ be a top level, sibling of _Frontend_ and _Backend_?

Comment (mdt2) [reply security/README.md]: Yes, good call!

File: security/backend/access_control/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+* Following the Positive Security Model principle, all pages must require login unless specifically designated as public.
+
+* Applications should use the default authentication for their platform whenever possible.
+
+* Authentication controls must be enforced on a trusted system. For example, on the server rather than the client.
+
+* Applications must use only HTTP POST to transmit credentials.
+
+* Passwords must be at least 8 characters in length and should incorporate two alternative characters.
```
Comment (cromwellryan) [inline security/backend/access_control/README.md:21]: I hesitate to include this. It is outdated guidance at this point.

For example, NIST has updated its [guidance](https://auth0.com/blog/dont-pass-on-the-new-nist-password-guidelines/). 

What about something like:

> Users should be encouraged and applications should support long, unique passwords.

File: security/backend/file_management/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# File Management
+
+**System configuration files must have restricted access.**
+
+This means
+
+1. They are not accessible in any way from the internet
+2. They are not accessible by insiders who have no need to access them
```
Comment (cromwellryan) [inline security/backend/file_management/README.md:8]: What is an insider?

Comment (mdt2) [reply security/backend/file_management/README.md:8]: An insider would be someone who's part of the organization or team

File: security/backend/file_management/README.md [hunk 4]
```diff
@@ -4,8 +4,8 @@
 
 This means
 
-1. They are not accessible in any way from the internet
-2. They are not accessible by insiders who have no need to access them
+1. They are not accessible in any way from the internet.
+2. They are not accessible by members of the organization or team who have no need to access them.
```
Comment (mdt2) [inline security/backend/file_management/README.md:8]: 👍

File: security/security_principles/README.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+## Principle of Least Privilege
+
+**A secure application gives the least amount of privilege required.**
+
+This principle encompasses user rights and resource permissions.
+
+For example, when we add clients to Slack, we only give them access to one external client channel.
+
+Another example is a server that only needs access to a network, read access to a database table, and the ability to write a log. So those are the only permissions that the server should be granted.
```
Comment (cromwellryan) [inline security/security_principles/README.md:41]:
```suggestion
Examples of least privilege include:

* only give clients access to one external client Slack channel
* only giving a server permission to a network, read access to a database table, and the ability to write a log
```

File: security/security_principles/README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
+# Security Principles for Developers
+
+Security on the web relies on confidentiality, integrity, and availability. Considering each of these key pieces while we build helps us produce robust security control. The following security principles embody these ideals.
```
Comment (cromwellryan) [inline security/security_principles/README.md:3]:
```suggestion
Security on the web relies on confidentiality, integrity, and availability. We are able to produce robust security control by considering each of these key pieces while we build software. The following security principles embody these ideals.
```

File: security/security_principles/README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+Keep in mind that it’s often easier to give additional permissions than to take away existing permissions.
+
+## Separation of Duties
+
+**A secure application separates duties, restricting the power of any individual.**
+
+Separation of duties manages conflict of interest by restricting any one individual’s power. It is best implemented by using role-based access control and by carefully defining what each role should be able to do.
+
+We can apply this method to job roles, where a developer develops, a QA tester tests, and a devops engineer deploys. This kind of separation ensures plenty of eyes on the code before it goes live.
```
Comment (cromwellryan) [inline security/security_principles/README.md:51]: While the example applies on paper, the title-to-responsibility alignment is anathema to our principles. Everyone tests to create shared responsibility.

What if we used environments instead:

```suggestion
We can apply this method to environments, where a developers would not have administrative access to production environments. This kind of separation ensures production changes only occur through controlled means.
```

File: security/security_principles/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+Another example of this method is how we write code at Sparkbox. We write modular, componentized code to limit complexity, which can limit what an attacker may be able to communicate with in an attack.
+
+## Security by Obscurity
+
+**A secure application is not reliant upon keeping details hidden.**
```
Comment (cromwellryan) [inline security/security_principles/README.md:57]: I _think_ we have this backwards. Security *by* obscurity means, don't tell folks what type of server or framework you are using. For example, it's best practice not to include the `X-Powered-By` header in HTTP repsonses.

File: security/security_principles/README.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+Security through obscurity means “don’t try to hide things.” Attackers have all the time in the world to find things we hide from them. Security by obscurity is a weak security control, meaning that systems shouldn’t be reliant upon keeping details hidden.
+
+For example, the security of a web application shouldn’t rely on knowledge of the source code being kept secret. That’s not to say keeping secrets is bad, but it’s not a reliable way to secure a web application on it’s own—be sure to use multiple levels of security on top of simply hiding information.
+
+## Do Not Trust the Client
+
+**A secure application validates information coming from the client.**
+
+This principle acknowledges that anything you send a client (a person or machine that’s consuming the application) can be changed and sent back. For this reason, validation should be performed in workflow or command patterns even if backend code performs validation as well. This validation process prevents unnecessary round trips to the backend and prevents the backend from revealing most tampered data.
```
Comment (cromwellryan) [inline security/security_principles/README.md:67]: The first statement is definitely accurate... what you send to the client can be tampered with. In fact, data sent to the server might not even come from your client code. (Postman is a good example of this).

For that reason, I think we can clarify this a little:

```suggestion
This principle acknowledges that anything you send a client (a person or machine that’s consuming the application) can be tampered with. Data sent from the client to the server should not be trusted. For this reason, the backend code should validate all data and permissions. 
```

I'm not sure I understand this.
> This validation process prevents unnecessary round trips to the backend and prevents the backend from revealing most tampered data.

**Review by cromwellryan — CHANGES_REQUESTED** [review]
**Overall**
I love the structure. Consumable and leaves room for the readers to explore more.

I have two types of comments here:
1. Some ways to clarify the language. This is my attempt at channeling Naomi 😄 
2. Correction/clarifications in areas that I think we have mistakes.

File: security/security_principles/README.md (hunk 9 above)
Comment (mdt2) [reply security/security_principles/README.md:67]: Yep this adjustment is way more clear 👍

File: security/security_principles/README.md (hunk 7 above)
Comment (mdt2) [reply security/security_principles/README.md:51]: TIL the word anathema. Great callout here.

File: security/security_principles/README.md (hunk 8 above)
Comment (mdt2) [reply security/security_principles/README.md:58]: Really good point. I made some adjustments, including your example. I also tried to include some of [Bill's points](https://docs.google.com/document/d/1tQpVG4X4nvxfCGRz4hqcg1NN2nWxvnZVEwTpv_WJfoI/edit#bookmark=id.6zs408ieahah) about security by obscurity potentially being a weak control

File: security/security_principles/README.md [hunk 10]
```diff
@@ -49,17 +49,17 @@ Keep in mind that it’s often easier to give additional permissions than to tak
...
 
 ## Security by Obscurity
 
 **A secure application is not reliant upon keeping details hidden.**
 
-Security through obscurity means “don’t try to hide things.” Attackers have all the time in the world to find things we hide from them. Security by obscurity is a weak security control, meaning that systems shouldn’t be reliant upon keeping details hidden.
+Security by obscurity means don't reveal what type of server or framework you're using. For example, it's best practice not to include the `X-Powered-By` header in HTTP responses.
 
-For example, the security of a web application shouldn’t rely on knowledge of the source code being kept secret. That’s not to say keeping secrets is bad, but it’s not a reliable way to secure a web application on it’s own—be sure to use multiple levels of security on top of simply hiding information.
+Security by obscurity can be a weak control if you're trying to hide information that you've shared. Attackers have all the time in the world to find things we hide from them, so systems shouldn’t be reliant only on keeping details hidden.
```
Comment (cromwellryan) [inline security/security_principles/README.md:62]: I like this!

File: security/security_principles/README.md [hunk 11]
```diff
@@ -49,17 +49,17 @@ Keep in mind that it’s often easier to give additional permissions than to tak
...
+We can apply this method to environments, where developers would not have administrative access to production environments. This kind of separation ensures production changes only occur through controlled means.
 
 Another example of this method is how we write code at Sparkbox. We write modular, componentized code to limit complexity, which can limit what an attacker may be able to communicate with in an attack.
 
 ## Security by Obscurity
 
 **A secure application is not reliant upon keeping details hidden.**
 
-Security through obscurity means “don’t try to hide things.” Attackers have all the time in the world to find things we hide from them. Security by obscurity is a weak security control, meaning that systems shouldn’t be reliant upon keeping details hidden.
+Security by obscurity means don't reveal what type of server or framework you're using. For example, it's best practice not to include the `X-Powered-By` header in HTTP responses.
```
Comment (cromwellryan) [inline security/security_principles/README.md:60]: How about this:

> Security through obscurity means limiting what is revealed about about the design or implementation of a system. For example, it's best practice not to include the `X-Powered-By` header in HTTP responses.

Comment (mdt2) [reply security/security_principles/README.md:60]: Yes! 💥

**Review by cromwellryan — APPROVED** [review]

**Pull request merged**